jaxopt/_src/iterative_refinement.py

```diff
@@ -0,0 +1,210 @@
+
+  The algorithm starts with :math:`r_0, x_0 = b, 0` and solves iteratively::
+
+    \begin{aligned}
+    \text{solve for x}  \bar{A} x &= r_{t-1}\\
+                        x_t &= x_{t-1} + x\\
+                        r_t &= b - A x_t
+    \end{aligned}
+  where :math:`\bar{A}` is some approximation of A, with preferably
+  better preconditonning than A.
+
+  This method has the advantage to converge even if the solve step is inaccurate.  
+  This is particularly useful for ill-posed problems, with strong regularization on
+  :math:`\bar{A}`, in ``float32`` environment, or with solvers that struggle to converge.
+
+  Attributes:
+    solve: a Callable that accepts A as first argument, b as second,
+      and a warm start ``init`` as third argument.
+      This solver can be inaccurate and run with low precision.
+    matvec: (optional) a Callable matvec_A(params_A, x).
+      By default, matvec_A(A, x) = tree_dot(A, x), where tree pytree A = params_A matches x structure.
+    matvec_A_bar: (optional) a Callable.
+      If None, then :math:`\bar{A}=A`. Otherwise, a Callable matvec_A_bar(x).
+    maxiter: maximum number of iterations (default: 10).
+    tol: absolute tolerance for stoping criterion (default: 1e-7).
+    verbose: If verbose=1, print error at each iteration.
+    implicit_diff: whether to enable implicit diff or autodiff of unrolled iterations.
+    implicit_diff_solve: the linear system solver to use.
+    jit: whether to JIT-compile the optimization loop (default: "auto").
+    unroll: whether to unroll the optimization loop (default: "auto")
+
+  References:
+    J. H. Wilkinson. Rounding Errors in Algebraic Processes. Prentice Hall, Englewood Cliffs, NJ, 1963.
+    Moler, C.B., 1967. Iterative refinement in floating point. Journal of the ACM (JACM), 14(2), pp.316-321.
+  """
+  matvec_A: Optional[Callable] = None
+  matvec_A_bar: Optional[Callable] = None
+  solve: Callable = partial(linear_solve.solve_gmres, ridge=1e-6)
+  maxiter: int = 10
+  tol: float = 1e-7
+  verbose: int = 0
+  implicit_diff_solve: Optional[Callable] = None
+  jit: base.AutoOrBoolean = "auto"
+  unroll: base.AutoOrBoolean = "auto"
+
+  def init_state(self,
+                 params: Any,
+                 params_A: Any,
+                 b: Any,
+                 params_A_bar: Any = None):
+    return IterativeRefinementState(
+      iter_num=0,
+      error=jnp.inf,
+      target_residuals=b,
+      init=params)
+
+  def update(self,
+             params: Any,
+             state: IterativeRefinementState,
+             params_A: Any,
+             b: Any,
+             params_A_bar: Optional[Any] = None):
+    if self._copy_params_A:
+      params_A_bar = params_A
+
+    A = self.matvec_A(params_A)
+    A_bar = self.matvec_A_bar(params_A_bar)
+
+    # TODO(lbethune): support preconditioners ?
+    # Could it be done by user with partial(solver, M=precond) ?
+    residual_sol = self.solve(A_bar, state.target_residuals, x0=state.init)
```

Comment:
In `cg`, the option is called `init` (https://github.com/google/jaxopt/blob/main/jaxopt/_src/linear_solve.py#L91). We should probably add an `init` option to `gmres` and other solvers instead of relying on `kwargs`...

Reply:
Good call !

On the other hand the `jax.scipy.sparse.linalg` interface (that the user might be more familar with) uses `x0`, and that's what is used under the hood for `jaxopt.linear_solve.solve_gmres`.

There is a mismatch between those two interfaces.